src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -2,26 +2,73 @@
 #include <sisl/logging/logging.h>
 #include <sisl/options/options.h>
 #include <sisl/settings/settings.hpp>
-#include "generated/resync_pg_shard_generated.h"
 #include "generated/resync_blob_data_generated.h"
+#include "generated/resync_pg_data_generated.h"
+#include "generated/resync_shard_data_generated.h"
+#include "replication_message.hpp"
+#include "lib/homeobject_impl.hpp"
+#include "hs_backend_config.hpp"
 
 namespace homeobject {
-
 HSHomeObject::PGBlobIterator::PGBlobIterator(HSHomeObject& home_obj, homestore::group_id_t group_id,
                                              uint64_t upto_lsn) :
-        home_obj_(home_obj), group_id_(group_id) {
+        home_obj_(home_obj), group_id_(group_id), snp_start_lsn_(upto_lsn) {
     auto pg = get_pg_metadata();
     pg_id_ = pg->pg_info_.id;
     repl_dev_ = static_cast< HS_PG* >(pg)->repl_dev_;
+    max_batch_size_ = HS_BACKEND_DYNAMIC_CONFIG(max_snapshot_batch_size_mb) * Mi;
+    if (max_batch_size_ == 0) { max_batch_size_ = DEFAULT_MAX_BATCH_SIZE_MB * Mi; }
+
     if (upto_lsn != 0) {
-        // Iterate all shards and its blob which have lsn <= upto_lsn
+        // Iterate all shards and its blobs which have lsn <= upto_lsn
         for (auto& shard : pg->shards_) {
-            auto sequence_num = home_obj_.get_sequence_num_from_shard_id(shard->info.id);
-            if (shard->info.lsn <= upto_lsn) { max_shard_seq_num_ = std::max(max_shard_seq_num_, sequence_num); }
+            if (shard->info.lsn <= upto_lsn) {
+                auto v_chunk_id = home_obj_.get_shard_v_chunk_id(shard->info.id);
+                shard_list_.emplace_back(shard->info, v_chunk_id.value());
+            }
         }
+        // sort shard list by <vchunkid, lsn> to ensure open shards positioned after sealed shards
+        std::ranges::sort(shard_list_, [](ShardEntry& a, ShardEntry& b) {
+            return a.v_chunk_num != b.v_chunk_num ? a.v_chunk_num < b.v_chunk_num : a.info.lsn < b.info.lsn;
```

Comment:
If I understood correctly, here assume shards with smaller v_chunk_num is more likely to be sealed. Could you pls explain the reason? (IIRC, shard is selected by available size)

Reply:
We don't care about the shard sequence across different chunks. This is just to ensure that **within each chunk** the opened shard is positioned last.

Let me edit the comment for clarity in the last push.